lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+        # type: (str) -> None
+        if self.ref_type not in self.known_types:
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types, self.ref_type))
+
+        if not isinstance(ref, six.string_types):
+            # This case can arise from user error when editing package.py files, so we
+            # want to catch this and handle it specially to provide more context.
+            raise InvalidGitRef('git reference was not a string: {0!r}'.format(ref))
+        self._ref = ref
+
+    def __repr__(self):
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self.ref_type, self._ref)
+
+    def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    @abc.abstractproperty
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """Return a hash prefix that the repo checkout should be compared against."""
+
+    @classmethod
+    def from_version_directive(cls, kwargs):
+        # type: (Dict) -> GitRef
+        # (1) Extract the relevant arguments.
+        commit = kwargs.pop('commit', None)
+        tag = kwargs.pop('tag', None)
+        branch = kwargs.pop('branch', None)
+        version_name = kwargs.pop('version_name', None)
+        # (2) Ensure no mutually exclusive kwargs are provided.
+        num_specified = len(list(filter(None, [commit, tag, branch])))
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
+            return cls.Branch(version_name)
+        # (4) If we are given a tag or a branch with a commit sha, note that by using
+        # a special constructor.
+        if num_specified == 2:
+            if tag and commit:
+                return cls.Tag(tag, commit=commit)
+            if branch and commit:
+                return cls.Branch(branch, commit=commit)
+        # (5) Otherwise, we have ambiguous ref arguments, or no ref arguments at all,
+        # so error.
+        if num_specified != 1:
+            raise InvalidGitRef(dedent("""\
+            Exactly one of 'tag' or 'branch' must be specified for a git url. 'commit'
+            may be provided in addition to either 'tag' or branch' to ensure that the
+            commit matches the given string.
+            Given:
+            commit={0}
+            tag={1}
+            branch={2}
+            """).format(commit, tag, branch))
+        if commit is not None:
+            return cls.Commit(commit)
+        if tag is not None:
+            return cls.Tag(tag)
+        assert branch is not None, (commit, tag, branch, version_name)
```

Comment:
Will this be weird for the user to see if the assertion fails? I guess they can just post a GitHub issue but maybe we could give them a more human friendly thing of how to fix it.

Reply:
The assertion should not fail because we have done the prior check in `__init__()` that we discussed here: https://github.com/spack/spack/pull/22235#discussion_r810546522. Does that make sense?